llama-index-core/llama_index/core/chat_engine/flow.py

```diff
@@ -0,0 +1,320 @@
+                )
+            prefix_messages = [
+                ChatMessage(content=system_prompt, role=llm.metadata.system_role)
+            ]
+
+        prefix_messages = prefix_messages or []
+        node_postprocessors = node_postprocessors or []
+
+        return cls(
+            retriever,
+            llm=llm,
+            memory=memory,
+            prefix_messages=prefix_messages,
+            node_postprocessors=node_postprocessors,
+            callback_manager=callback_manager_from_settings_or_context(
+                Settings, service_context
+            ),
+            text_qa_template=text_qa_template,
+        )
+
+    def _generate_qa_context(self, message: str) -> Tuple[str, List[NodeWithScore]]:
+        """Generate context information from a message."""
+        nodes = self._retriever.retrieve(message)
+        for postprocessor in self._node_postprocessors:
+            nodes = postprocessor.postprocess_nodes(
+                nodes, query_bundle=QueryBundle(message)
+            )
+
+        context_str = "\n\n".join(
+            [n.node.get_content(metadata_mode=MetadataMode.LLM).strip() for n in nodes]
+        )
+
+        return (
+            self._text_qa_template.format(context_str=context_str, query_str=message),
+            nodes,
+        )
+
+    async def _agenerate_qa_context(
+        self, message: str
+    ) -> Tuple[str, List[NodeWithScore]]:
+        """Generate context information from a message."""
+        nodes = await self._retriever.aretrieve(message)
+        for postprocessor in self._node_postprocessors:
+            nodes = postprocessor.postprocess_nodes(
+                nodes, query_bundle=QueryBundle(message)
+            )
+        context_str = "\n\n".join(
+            [n.node.get_content(metadata_mode=MetadataMode.LLM).strip() for n in nodes]
+        )
+
+        return (
+            self._context_template.format(context_str=context_str, query_str=message),
+            nodes,
+        )
+
+    @trace_method("chat")
+    def chat(
+        self, message: str, chat_history: Optional[List[ChatMessage]] = None
+    ) -> AgentChatResponse:
+        if chat_history is not None:
+            self._memory.set(chat_history)
+
+        qa_context_str_template, nodes = self._generate_qa_context(message)
+
+        user_message = ChatMessage(
+            role=MessageRole.USER, content=qa_context_str_template
+        )
+        self._memory.put(user_message)
+
+        prefix_messages_token_count = len(
+            self._memory.tokenizer_fn(
+                " ".join([(m.content or "") for m in self._prefix_messages])
+            )
+        )
+        all_messages = self._prefix_messages + self._memory.get(
+            initial_token_count=prefix_messages_token_count
+        )
+        chat_response = self._llm.chat(all_messages)
+        ai_message = chat_response.message
+        self._memory.put(ai_message)
+
+        self._memory.get_all()[-2].content = message
```

Comment:
This feels very... weird? What is happening here? Can this be done before putting things in memory? If not, this needs a comment explaining